### DISCUSSION TITLE
Configure Pino.js to work with pm2

### DISCUSSION THREAD

**Discussion by favoyang** [verdaccio/verdaccio#3415] [Q&A]
**Describe the bug**

Pino.js has an issue with pm2 if you manually set the destination to descriptor 1. Because pm2 overwrites `process.stdout.write` with its own function. see https://github.com/pinojs/pino/issues/491#issuecomment-414075492

Pino fixed this issue in https://github.com/pinojs/pino/pull/493/files if you don't specify the destination when writing to stdout.

https://github.com/pinojs/pino/issues/491

**To Reproduce**

See this comment, https://github.com/pinojs/pino/issues/491#issuecomment-414043388

**Expected behavior**

Let pino.js handle the magic by removing the destination.

```
  const logger = pino(pinoConfig, destination);
=>
  const logger = pino(pinoConfig);
```

**Configuration File (cat ~/.config/verdaccio/config.yaml)**

```
logs: { type: stdout, format: pretty, level: http }
```

**Environment information**

```
verdaccio: 5.x
```

**Comment by juanpicado** [comment]
Thanks @favoyang for sharing it 💯

**Comment by amitgilad3** [comment]
Hey, so i'm working on a fix, and i was wondering if we want to remove the destination from the logger in case of stdout ?
any suggestions  @juanpicado ??

**Comment by juanpicado** [comment]
> Hey, so i'm working on a fix, and i was wondering if we want to remove the destination from the logger in case of stdout ?
> 
> any suggestions  @juanpicado ??

I was trying a bit this afternoon removing the default destination but I broke all test and did not continue. Go ahead if you want to give a try, but yeah that's the idea.

**Comment by amitgilad3** [comment]
Hey @favoyang , I working on this issue and was wondering if you can provide steps to reproduce.
I am running verdaccio with pm2 and I see the logs.

**Comment by favoyang** [comment]
Create directory: /srv/verdaccio

Add package.json, run `npm install` to install verdaccio 5.0.4
```
{
  "name": "verdaccio-server",
  "dependencies": {
    "verdaccio": "^5.0.4"
  }
}
```

Install pm2
```
npm install pm2 -g
```

Add pm2-verdaccio.yml
```
apps:
  - name: verdaccio
    cwd: /srv/verdaccio
    script: ./node_modules/.bin/verdaccio
    exec_mode: cluster
    instances: 1
    error_file: ./verdaccio-error.log
    out_file: ./verdaccio.log
    combine_logs: true
    max_memory_restart: 2G
    max_restarts: 5
    autorestart: true
```

Run pm2 with the pm2-verdaccio.yml configuration file.
```
$ pm2 start pm2-verdaccio.yml
[PM2][WARN] Applications verdaccio not running, starting...
[PM2] App [verdaccio] launched (1 instances)
┌─────┬──────────────┬─────────────┬─────────┬─────────┬──────────┬────────┬──────┬───────────┬──────────┬──────────┬──────────┬──────────┐
│ id  │ name         │ namespace   │ version │ mode    │ pid      │ uptime │ ↺    │ status    │ cpu      │ mem      │ user     │ watching │
├─────┼──────────────┼─────────────┼─────────┼─────────┼──────────┼────────┼──────┼───────────┼──────────┼──────────┼──────────┼──────────┤
│ 0   │ verdaccio    │ default     │ 1.0.0   │ cluster │ 15834    │ 0s     │ 0    │ online    │ 0%       │ 34.1mb   │ favo     │ disabled │
└─────┴──────────────┴─────────────┴─────────┴─────────┴──────────┴────────┴──────┴───────────┴──────────┴──────────┴──────────┴──────────┘
```

Test the verdaccio server
```
$ curl -s http://localhost:4873/-/all
{"_updated":99999}
```

Check pm2 logs. Notice that logs wrote to $home/.pm2/pm2.log instead of /srv/verdaccio/verdaccio.log (defined in pm2 configuration file), caused by the reason described in the first post. This reproduced the issue.

```
$ pm2 logs
[TAILING] Tailing last 15 lines for [all] processes (change the value with --lines option)
/home/favo/.pm2/pm2.log last 15 lines:
PM2        | 2021-04-29T21:43:18: PM2 log: Process dump file    : /home/favo/.pm2/dump.pm2
PM2        | 2021-04-29T21:43:18: PM2 log: Concurrent actions   : 2
PM2        | 2021-04-29T21:43:18: PM2 log: SIGTERM timeout      : 1600
PM2        | 2021-04-29T21:43:18: PM2 log: ===============================================================================
PM2        | 2021-04-29T21:44:12: PM2 log: App [verdaccio:0] starting in -cluster mode-
PM2        | 2021-04-29T21:44:12: PM2 log: App [verdaccio:0] online
PM2        |  warn --- config file  - /home/favo/.config/verdaccio/config.yaml
PM2        |  warn --- Plugin successfully loaded: verdaccio-htpasswd
PM2        |  warn --- Plugin successfully loaded: verdaccio-audit
PM2        |  warn --- http address - http://localhost:4873/ - verdaccio/5.0.4
PM2        |  http --- 127.0.0.1 requested 'GET /'
PM2        |  http --- 200, user: null(127.0.0.1), req: 'GET /', bytes: 0/1023
PM2        |  http --- 127.0.0.1 requested 'GET /-/all'
PM2        |  info --- making request: 'GET https://registry.npmjs.org/-/all'
PM2        |  http --- 200, user: null(127.0.0.1), req: 'GET /-/all', bytes: 0/19

/srv/verdaccio/verdaccio.log last 15 lines:
/srv/verdaccio/verdaccio-error.log last 15 lines:
```

Now edit ./node_modules/verdaccio/build/lib/logger/logger.js, (just a quick hack to prove the solution because verdaccio still needs the destination for the file log).

```
line 72  const logger = (0, _pino.default)(pinoConfig, destination);
=>
const logger = (0, _pino.default)(pinoConfig);
```

Restart server and touch it
```
pm2 restart pm2-verdaccio.yml
curl -s http://localhost:4873/-/all
```

Check logs. Now the logs wrote to the verdaccio.log file.
```
$ pm2 logs
[TAILING] Tailing last 15 lines for [all] processes (change the value with --lines option)
/home/favo/.pm2/pm2.log last 15 lines:
...
PM2        | 2021-04-29T21:56:20: PM2 log: Stopping app:verdaccio id:0
PM2        | 2021-04-29T21:56:20: PM2 log: App name:verdaccio id:0 disconnected
PM2        | 2021-04-29T21:56:20: PM2 log: App [verdaccio:0] exited with code [0] via signal [SIGINT]
PM2        | 2021-04-29T21:56:20: PM2 log: pid=17207 msg=process killed
PM2        | 2021-04-29T21:56:20: PM2 log: App [verdaccio:0] starting in -cluster mode-
PM2        | 2021-04-29T21:56:20: PM2 log: App [verdaccio:0] online

/srv/verdaccio/verdaccio-error.log last 15 lines:
/srv/verdaccio/verdaccio.log last 15 lines:
0|verdacci |  warn --- config file  - /home/favo/.config/verdaccio/config.yaml
0|verdacci |  warn --- Plugin successfully loaded: verdaccio-htpasswd
0|verdacci |  warn --- Plugin successfully loaded: verdaccio-audit
0|verdacci |  warn --- http address - http://localhost:4873/ - verdaccio/5.0.4
0|verdacci |  http --- 127.0.0.1 requested 'GET /-/all'
0|verdacci |  info --- making request: 'GET https://registry.npmjs.org/-/all'
0|verdacci |  http --- 200, user: null(127.0.0.1), req: 'GET /-/all', bytes: 0/19
```

**Comment by amitgilad3** [comment]
Thanks @favoyang  for the info, i already have a fix in place.

**Comment by stale** [comment]
Hi pal 👋🏼!
This issue has gone quiet 😶.

 We get a lot of issues, so we currently close issues after 180 days of inactivity. It’s been at least 90c days since the last update here.
 If we missed this issue or if you want to keep it open, please reply here. You can also add/suggest the label "discuss" to keep this issue open!
 As a friendly reminder: the best way to see this issue, or any other, fixed is to open a Pull Request. Check out [https://github.com/verdaccio/contributing](https://github.com/verdaccio/contributing) for more information about opening PRs, triaging issues, and contributing!

 Thanks for being a part of the Verdaccio community! 💘
